Feature/275 masstransit rabbitmq client integration

MichielBrys opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠️ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! -->

<!-- ⚠️ **Every PR** needs to have a linked issue and have previously been approved. PRs that don't follow this will be rejected. >

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Closes #275 **

<!-- Add an overview of the changes here -->

Host & client projects for masstransit rabbitmq. And example project. Readme's are in the src projects.

<!-- All details should be in the linked issue. Feel free to call out any outstanding differences here. -->

## PR Checklist

<!-- Please check if your PR fulfills the following requirements, and remove the ones that are not applicable to the current PR -->

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] New integration
  - [x] Docs are written
  - [x] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

## Other information

<!-- Please add any other information that might be helpful to reviewers. -->

MichielBrys commented:
> Contributor @dotnet-policy-service agree

fredimachado commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitClientExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+            throw new ArgumentException("The name must be a valid, non-empty string.", nameof(name));
+
+        var configuration = services.BuildServiceProvider().GetRequiredService<IConfiguration>();
+        var configurationSection = configuration.GetSection($"MassTransit:{name}");
```
I believe the section (according to the other Aspire clients) should be `Aspire:MassTransit:Client`

fredimachado replied:
I've done something like this in my PR as well, which also allows to bind a section with the `name`.
https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-bfd900a6101f7a145a030e21810fac284c3397f522f7adeb3ade6c2bb976ad6bR59

fredimachado commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitClientExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+/// <summary>
+/// Configuration options for MassTransit client integration 
+/// </summary>
+internal sealed class MassTransitOptions
```
This could be a public `MassTransitClientSettings` with XML docs and a `DisableTelemetry` prop.

Instead of having a `telemetry` parameter in `AddMassTransitClient` extension method, you could have an `Action<MassTransitClientSettings>` typed parameter to allow set up of some values after binding the config.

fredimachado commented:
Hi @MichielBrys , I noticed the `AddMassTransitClient` and I was just wondering what would happen if, in the future we also have Aspire packages for other transports like Service Bus for example.

Maybe it could be named `AddMassTransitRabbitMqClient` in order to not make it ambiguous?

aaronpowell commented on .idea/.idea.CommunityToolkit.Aspire/.idea/.gitignore:
Can you make sure the `.idea` folder is in the git ignore and these files aren't committed.

aaronpowell commented on CommunityToolkit.Aspire.sln:
Client integrations don't require `Client` as part of the name

aaronpowell commented on examples/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.ApiService/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.ApiService.csproj:
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+    <PropertyGroup>
+        <TargetFramework>net8.0</TargetFramework>
```
Remove TFM, we control that higher up

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost.csproj:
```diff
@@ -0,0 +1,23 @@
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Aspire.Hosting.AppHost" Version="9.0.0" />
```
```suggestion
        <PackageReference Include="Aspire.Hosting.AppHost" />
```

NuGet [CPM](https://learn.microsoft.com/en-us/nuget/consume-packages/Central-Package-Management) will set the version of the package

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost.csproj:
```diff
@@ -0,0 +1,23 @@
+
+    <ItemGroup>
+        <PackageReference Include="Aspire.Hosting.AppHost" Version="9.0.0" />
+
+    </ItemGroup>
```
```suggestion
    </ItemGroup>
```

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost.csproj:
```diff
@@ -0,0 +1,23 @@
+    <ItemGroup>
+      <ProjectReference Include="..\CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.ApiService\CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.ApiService.csproj" />
+      <ProjectReference Include="..\..\src\CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ\CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj" IsAspireProjectResource="false" />
+
+    </ItemGroup>
```
```suggestion
    </ItemGroup>
```

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost/Program.cs:
```diff
@@ -0,0 +1,19 @@
+
+using CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ;
```
```suggestion
using CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ;
```

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.ServiceDefaults/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.ServiceDefaults.csproj:
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net8.0</TargetFramework>
```
Remove TFM, we control that higher up

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,18 @@
+    <PropertyGroup>
+      <AdditionalPackageTags>Client Aspire MassTransit RabbitMQ</AdditionalPackageTags>
+      <Description>A .NET Aspire client integration for MassTransit RabbitMQ.</Description>
+        <TargetFramework>net8.0</TargetFramework>
```
Remove TFM, we control that higher up

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,18 @@
+      <AdditionalPackageTags>Client Aspire MassTransit RabbitMQ</AdditionalPackageTags>
+      <Description>A .NET Aspire client integration for MassTransit RabbitMQ.</Description>
+        <TargetFramework>net8.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
```
```suggestion
      <ImplicitUsings>enable</ImplicitUsings>
      <Nullable>enable</Nullable>
    </PropertyGroup>
```

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,18 @@
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+
+  <ItemGroup>
```
```suggestion

  <ItemGroup>
```

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,18 @@
+
+
+  <ItemGroup>
+    <PackageReference Include="MassTransit.RabbitMQ" Version="8.3.2"/>
+    <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.10.0"/>
```
```suggestion
    <PackageReference Include="MassTransit.RabbitMQ" />
    <PackageReference Include="OpenTelemetry.Extensions.Hosting" />
```

NuGet CPM will handle versions

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,18 @@
+    <PackageReference Include="MassTransit.RabbitMQ" Version="8.3.2"/>
+    <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.10.0"/>
+  </ItemGroup>
+
+  
+</Project>
```
```suggestion
</Project>
```

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitClientExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+using Microsoft.Extensions.DependencyInjection;
+using System.Reflection;
+
+namespace CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ;
```
The namespace for a client integration extension should be `Microsoft.Extensions.Hosting`

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitClientExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+    /// <param name="services">The client service collection.</param>
+    /// <param name="name">A unique name for the RabbitMQ instance.</param>
+    /// <param name="telemetry">Enables telemetry, which could be exported to either OpenTelemetry or Application Insights.</param>
+    public static void AddMassTransitClient(this IServiceCollection services, string name, bool? telemetry = false)
```
Client integrations should be extensions of `IHostApplicationBuilder` not `IServiceCollection`

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitClientExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+    /// <param name="telemetry">Enables telemetry, which could be exported to either OpenTelemetry or Application Insights.</param>
+    public static void AddMassTransitClient(this IServiceCollection services, string name, bool? telemetry = false)
+    {
+        if (string.IsNullOrWhiteSpace(name))
+            throw new ArgumentException("The name must be a valid, non-empty string.", nameof(name));
```
```suggestion
ArgumentException.ThrowIfNullOrWhiteSpace(name, nameof(name));
```

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitClientExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+            });
+        });
+
+        if (telemetry == true)
```
```suggestion
        if (telemetry)
```

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitClientExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+
+        var configuration = services.BuildServiceProvider().GetRequiredService<IConfiguration>();
+        var configurationSection = configuration.GetSection($"MassTransit:{name}");
+        var options = new MassTransitOptions();
```
For options management, we need to provide an `Action` callback that can be optionally provided to perform configuration in code rather than via config: https://github.com/CommunityToolkit/Aspire/blob/main/src/CommunityToolkit.Aspire.OllamaSharp/AspireOllamaSharpExtensions.cs#L40

Check out how https://github.com/CommunityToolkit/Aspire/blob/main/src/CommunityToolkit.Aspire.OllamaSharp/AspireOllamaSharpExtensions.cs is implemented for more insights.

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/README.md:
````diff
@@ -0,0 +1,27 @@
+
+### Example Usage
+
+```csharp
+builder.Services.AddMassTransitClient("RabbitMQ", telemetry: true);
````
Missing an end to the code fence

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
```
```suggestion
﻿<Project Sdk="Microsoft.NET.Sdk">
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,17 @@
+    <PropertyGroup>
+      <AdditionalPackageTags>Hosting Aspire MassTransit RabbitMQ</AdditionalPackageTags>
+      <Description>A .NET Aspire integration for hosting MassTransit RabbitMQ services.</Description>
+      <TargetFramework>net8.0</TargetFramework>
```
Remove TFM, we control that higher up

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,17 @@
+      <AdditionalPackageTags>Hosting Aspire MassTransit RabbitMQ</AdditionalPackageTags>
+      <Description>A .NET Aspire integration for hosting MassTransit RabbitMQ services.</Description>
+      <TargetFramework>net8.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
```
```suggestion
       <ImplicitUsings>enable</ImplicitUsings>
       <Nullable>enable</Nullable>
    </PropertyGroup>
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,17 @@
+
+  <ItemGroup>
+    <PackageReference Include="Aspire.Hosting.RabbitMQ" Version="8.2.2"/>
+
+  </ItemGroup>
+
+
```
```suggestion
  </ItemGroup>
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/MassTransitRabbitMQHostingExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+using Aspire.Hosting.ApplicationModel;
+using Microsoft.Extensions.Configuration;
+
+namespace CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ;
```
Namespace for hosting integrations should use `Aspire.Hosting`

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/MassTransitRabbitMQHostingExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+    public static IResourceBuilder<RabbitMQServerResource> AddMassTransit(this IDistributedApplicationBuilder builder,
+        string name, Action<MassTransitOptions>? configure = null)
+    {
+        if (string.IsNullOrWhiteSpace(name))
+            throw new ArgumentException("The name must be a valid, non-empty string.", nameof(name));
```
```suggestion
        ArgumentNullException.ThrowIfNull(builder, nameof(builder));
        ArgumentNullException.ThrowIfNull(name, nameof(name));
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/MassTransitRabbitMQHostingExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+    /// <param name="builder">The application builder.</param>
+    /// <param name="name">A unique name for the RabbitMQ instance.</param>
+    /// <param name="configure">An optional configuration action to override default settings.</param>
+    public static IResourceBuilder<RabbitMQServerResource> AddMassTransit(this IDistributedApplicationBuilder builder,
+        string name, Action<MassTransitOptions>? configure = null)
```
```suggestion
    public static IResourceBuilder<RabbitMQServerResource> AddMassTransit(this IDistributedApplicationBuilder builder, [ResourceName] string name, Action<MassTransitOptions>? configure = null)
```

Ensure that we have the `ResourceName` attribute on the name parameter for a resource.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/MassTransitRabbitMQHostingExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+            throw new ArgumentException("The name must be a valid, non-empty string.", nameof(name));
+
+        // Load options from configuration
+        IConfigurationSection configurationSection = builder.Configuration.GetSection($"MassTransit:{name}");
```
Settings keys should be prefixed with `Aspire`

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/MassTransitRabbitMQHostingExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+    /// <summary>
+    /// Username to use for RabbitMQ.
+    /// </summary>
+    public string Username { get; set; } = string.Empty;
```
This, and `Password` should be ParameterResource types so that we can ensure we're securely managing them. Have a review of https://learn.microsoft.com/dotnet/aspire/extensibility/secure-communication-between-integrations to understand the pattern more.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/MassTransitRabbitMQHostingExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+/// <summary>
+/// Configuration options for MassTransit integration (same as host-side).
+/// </summary>
+public sealed class MassTransitOptions
```
Type should be in its own file.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/README.md:
````diff
@@ -0,0 +1,32 @@
+
+### Example Usage
+
+```csharp
````
Missing a closing code fence

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,17 @@
+    </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Aspire.Hosting.RabbitMQ" Version="8.2.2"/>
```
```suggestion
    <PackageReference Include="Aspire.Hosting.RabbitMQ" />
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,17 @@
+
+  <ItemGroup>
+    <PackageReference Include="Aspire.Hosting.RabbitMQ" Version="8.2.2"/>
+
+  </ItemGroup>
```
```suggestion
  </ItemGroup>
```

aaronpowell reviewed (changes requested):
Left some comments and suggestions to bring it inline with more of the design of the Toolkit projects.

Also, you'll need to ensure that the NuGet packages are added to the `Directory.Packages.props` with their correct versions since we use CPM to ensure everything is consistently versioned.

Alirexaa commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/MassTransitRabbitMQHostingExtensions.cs:
```diff
@@ -0,0 +1,50 @@
+/// <summary>
+/// MassTransitHostingExtensions provides extension methods for configuring MassTransit in an Aspire-based application.
+/// </summary>
+public static class MassTransitRabbitMqHostingExtensions
```
I do not see a reason for having hosting integration for MassTransit; it's not hosting.

We should use the connection string provided from normal `AddRabbitMQ("rabbitmq")` for creating client integration.

fredimachado replied:
That's a great point. Having just a client integration makes more sense for this.

aaronpowell replied:
Agreed - looking at the implementation of the integration, it is wrapping the RabbitMQ integration and doesn't seem to add any additional functionality.

anoordover commented:
Can you explain what this solution adds on the ease of use as compared to https://github.com/anoordover/MassTransitJobConsumer?
What would the impact of using this component be for my example?

anoordover commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitRabbitMqExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+        ArgumentException.ThrowIfNullOrWhiteSpace(name, nameof(name));
+
+        // Load options from configuration
+        var serviceProvider = builder.Services.BuildServiceProvider();
```
The serviceprovider is an IDisposable. Shouldn't we dispose this?

aaronpowell replied:
We shouldn't be resolving a service provider within this phase. To access configuration, use `builder.Configuration`: https://github.com/CommunityToolkit/Aspire/blob/main/src/CommunityToolkit.Aspire.OllamaSharp/AspireOllamaSharpExtensions.cs#L106-L109

anoordover commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/MassTransitRabbitMqExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+                .WithMetrics(b => b.AddMeter(InstrumentationOptions.MeterName))
+                .WithTracing(providerBuilder =>
+                {
+                    providerBuilder.AddSource(InstrumentationOptions.MeterName);
```
According to the MassTransit documentation you should use DiagnosticHeaders.DefaultListenerName for the source. See https://masstransit.io/documentation/configuration/observability#aspnet-core-application.

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.ServiceDefaults/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.ServiceDefaults.csproj:
```diff
@@ -0,0 +1,21 @@
+    <ItemGroup>
+        <FrameworkReference Include="Microsoft.AspNetCore.App"/>
+
+        <PackageReference Include="Microsoft.Extensions.Http.Resilience" Version="8.10.0"/>
+        <PackageReference Include="Microsoft.Extensions.ServiceDiscovery" Version="8.2.2"/>
+        <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.9.0"/>
+        <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.9.0"/>
+        <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.9.0"/>
+        <PackageReference Include="OpenTelemetry.Instrumentation.Http" Version="1.9.0"/>
+        <PackageReference Include="OpenTelemetry.Instrumentation.Runtime" Version="1.9.0"/>
```
Please ensure packages are added via NuGet tooling to use the package version defined in `Directory.Packages.props`.

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+      <AdditionalPackageTags>Client Aspire MassTransit RabbitMQ</AdditionalPackageTags>
```
```suggestion
      <AdditionalPackageTags>Client MassTransit RabbitMQ</AdditionalPackageTags>
```

Aspire isn't required as a custom tag

aaronpowell commented on src/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,14 @@
+    </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="MassTransit.RabbitMQ" Version="8.3.2" />
```
Please ensure packages are added using NuGet tooling and use the version defined in `Directory.Package.props`

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,14 @@
+    </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Aspire.Hosting.RabbitMQ" Version="9.0.0"/>
```
Please ensure packages are added using NuGet tooling so the package version is registered in the `Directory.Package.props`.

Also, for Aspire packages, use the variable available for the version.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+      <AdditionalPackageTags>Hosting Aspire MassTransit RabbitMQ</AdditionalPackageTags>
```
```suggestion
      <AdditionalPackageTags>Hosting MassTransit RabbitMQ</AdditionalPackageTags>
```

Aspire isn't required as a tag

MichielBrys commented:
Thanks for the reviews, I read them and indeed realised the hosting package isn't doing anything extra so I got rid of it. Currently only have the client which uses the connection string from the .AddRabbitMQ().

I also fixed the nuget versioning to use the Directory.Packages.props file.

aaronpowell commented on Directory.Packages.props:
```diff
@@ -8,9 +8,11 @@
     <PackageVersion Include="Aspire.Hosting.SqlServer" Version="$(AspireVersion)" />
     <!-- AspNetCore packages -->
     <PackageVersion Include="AspNetCore.HealthChecks.SqlServer" Version="8.0.2" />
+    <PackageVersion Include="MassTransit.RabbitMQ" Version="8.3.2" />
```
Can you move this down under the **External Packages** section

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.AppHost/appsettings.json:
```diff
@@ -0,0 +1,13 @@
+      "Microsoft.AspNetCore": "Warning",
+      "Aspire.Hosting.Dcp": "Warning"
+    }
+  },
+  "Parameters": {
+    "RabbitMQUsername": "JgmewmgMGIOghehomiH4988",
+    "RabbitMQPassword": "wrhHOmhmiemIGMJOSDIOMHMOIQji15s1g5e6156e"
+  }
```
```suggestion
  }
```

Even though they are dummy values, let's not include them in source control

aaronpowell commented on examples/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.ServiceDefaults/CommunityToolkit.Aspire.Hosting.MassTransit.RabbitMQ.ServiceDefaults.csproj:
```diff
@@ -0,0 +1,23 @@
+        <IsAspireSharedProject>true</IsAspireSharedProject>
+    </PropertyGroup>
+
+    <ItemGroup>
+
+    </ItemGroup>
+
```
```suggestion
```

`ItemGroup` not required.

aaronpowell commented on src/CommunityToolkit.Aspire.MassTransit.RabbitMQ/CommunityToolkit.Aspire.MassTransit.RabbitMQ.csproj:
```diff
@@ -0,0 +1,18 @@
+      <Nullable>enable</Nullable>
+      <RootNamespace>CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ</RootNamespace>
+    </PropertyGroup>
+
+  <ItemGroup>
+    </ItemGroup>
+
```
```suggestion

```

aaronpowell commented on src/CommunityToolkit.Aspire.MassTransit.RabbitMQ/MassTransitRabbitMqOptions.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ;
```
```suggestion
﻿namespace CommunityToolkit.Aspire.MassTransit.RabbitMQ;
```

aaronpowell commented on src/CommunityToolkit.Aspire.MassTransit.RabbitMQ/PublicAPI.Shipped.txt:
```diff
@@ -0,0 +1,7 @@
+#nullable enable
+CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.MassTransitRabbitMqOptions
+CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.MassTransitRabbitMqOptions.DisableTelemetry.get -> bool
+CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.MassTransitRabbitMqOptions.DisableTelemetry.set -> void
+CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.MassTransitRabbitMqOptions.MassTransitRabbitMqOptions() -> void
+Microsoft.Extensions.Hosting.MassTransitRabbitMqExtensions
+static Microsoft.Extensions.Hosting.MassTransitRabbitMqExtensions.AddMassTransitRabbitMq(this Microsoft.Extensions.Hosting.IHostApplicationBuilder! builder, string! name, System.Action<CommunityToolkit.Aspire.Client.MassTransit.RabbitMQ.MassTransitRabbitMqOptions!>? configure = null) -> void
```
```suggestion
CommunityToolkit.Aspire.MassTransit.RabbitMQ.MassTransitRabbitMqOptions
CommunityToolkit.Aspire.MassTransit.RabbitMQ.MassTransitRabbitMqOptions.DisableTelemetry.get -> bool
CommunityToolkit.Aspire.MassTransit.RabbitMQ.MassTransitRabbitMqOptions.DisableTelemetry.set -> void
CommunityToolkit.Aspire.MassTransit.RabbitMQ.MassTransitRabbitMqOptions.MassTransitRabbitMqOptions() -> void
Microsoft.Extensions.Hosting.MassTransitRabbitMqExtensions
static Microsoft.Extensions.Hosting.MassTransitRabbitMqExtensions.AddMassTransitRabbitMq(this Microsoft.Extensions.Hosting.IHostApplicationBuilder! builder, string! name, System.Action<CommunityToolkit.Aspire.MassTransit.RabbitMQ.MassTransitRabbitMqOptions!>? configure = null) -> void
```

aaronpowell commented:
> Thanks for the reviews, I read them and indeed realised the hosting package isn't doing anything extra so I got rid of it. Currently only have the client which uses the connection string from the .AddRabbitMQ().
> 
> I'm thinking it would be pretty usefull to have in the community toolkit and I'd like to expand it to support all the implementations that work with masstransit so people have a nice wrapper to use with aspire to easily set up masstransit
> 
> I also fixed the nuget versioning to use the Directory.Packages.props file.

Given that it doesn't provide any functionality over the standard RabbitMQ integration, I don't see that there's a need for the hosting integration.

The client integration should be able to use the connection string provided by the standard RabbitMQ integration just fine.

MichielBrys commented:
> > Thanks for the reviews, I read them and indeed realised the hosting package isn't doing anything extra so I got rid of it. Currently only have the client which uses the connection string from the .AddRabbitMQ().
> > I'm thinking it would be pretty usefull to have in the community toolkit and I'd like to expand it to support all the implementations that work with masstransit so people have a nice wrapper to use with aspire to easily set up masstransit
> > I also fixed the nuget versioning to use the Directory.Packages.props file.
> 
> Given that it doesn't provide any functionality over the standard RabbitMQ integration, I don't see that there's a need for the hosting integration.
> 
> The client integration should be able to use the connection string provided by the standard RabbitMQ integration just fine.

Agreed its the client package I meant that could be nice to have in the toolkit, the hosting one I already removed yesterday.

aaronpowell commented:
Also, can the examples be moved into a sub-folder of `examples`, such as `examples/masstransit-rabbitmq` (and a solution folder to match), so that it's consistent with the others.

aaronpowell commented:
> Could someone merge the pr? I don't have write access I think. Thanks :)

Yeah. CodeQL didn't run for some reason so I'll have to override that to merge.

aaronpowell reviewed (changes requested):
I just noticed that there's no tests.

Can you add some tests for the integration. Since this is a client integration, check out https://github.com/CommunityToolkit/Aspire/tree/main/tests/CommunityToolkit.Aspire.Meilisearch.Tests as an example for how to write client integration tests.

aaronpowell commented:
I merged locally (wanted to do a few bits of tidying up) and it looks like it's freaked out this PR a little, so I'll close the PR as everything is already in the `main` branch.

Should be live on NuGet soon.
